wallet/src/de/schildbach/wallet/ui/backup/BackupWalletDialogFragment.java

```diff
@@ -302,21 +342,106 @@ private void handleGo() {
         final String passwordAgain = passwordAgainView.getText().toString().trim();
 
         if (passwordAgain.equals(password)) {
-            backupWallet();
+            final Wallet wallet = walletActivityViewModel.wallet.getValue();
+            if (wallet.isEncrypted()) {
+                setState(BackupWalletViewModel.State.CRYPTING);
+                executor.execute(() -> {
+                    final String inputPIN = spendingPINView.getText().toString().trim();
+
+                    try {
+                        backupWallet = new WalletProtobufSerializer()
+                                .readWallet(Constants.NETWORK_PARAMETERS, null,
+                                        new WalletProtobufSerializer().walletToProto(wallet));
+
+                        final KeyCrypter keyCrypter = backupWallet.getKeyCrypter();
+                        final KeyParameter derivedAesKey = keyCrypter.deriveKey(inputPIN);
+                        backupWallet.decrypt(derivedAesKey);
+                        log.info("backup wallet successfully decrypted for back up");
+                        setState(BackupWalletViewModel.State.EXPORTING);
+                        backupWallet();
+
+                    } catch (final Wallet.BadWalletEncryptionKeyException e) {
+                        log.info("wallet decryption failed, bad spending password: " + e.getMessage());
+                        backupWallet = null;
+                        setState(BackupWalletViewModel.State.BADPIN);
+                    } catch (UnreadableWalletException e) {
+                        log.info("wallet deserialization failed: " + e.getMessage());
+                        ErrorDialogFragment.showDialog(getParentFragmentManager(), e.toString());
+                    }
+                });
+            } else {
+                setState(BackupWalletViewModel.State.EXPORTING);
+                backupWallet = walletActivityViewModel.wallet.getValue();
+                backupWallet();
+                backupWallet = null;
```

Comment:
Not sure if the removal of the reference to the decrypted wallet is needed in general. We need to trust the garbage collector anyway. If you prefer to keep it, I'd move it towards the end of the `CreateDocument` callback (around line 171) because that's where we're done with it.